// @bendera/vscode-webview-elements/dist/vscode-checkbox.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { html, nothing } from 'lit';
import { customElement, property, state } from 'lit/decorators.js';
import { classMap } from 'lit/directives/class-map.js';
import { FormButtonWidgetBase } from './includes/form-button-widget/FormButtonWidgetBase';
import baseStyles from './includes/form-button-widget/base.styles';
import checkboxStyles from './includes/form-button-widget/checkbox.styles';
import formHelperTextStyles from './includes/formHelperTextStyles';
/**
 * @attr name - Name which is used as a variable name in the data of the form-container.
 */
let VscodeCheckbox = class VscodeCheckbox extends FormButtonWidgetBase {
    constructor() {
        super(...arguments);
        this.checked = false;
        this.label = '';
        this.value = '';
        this.disabled = false;
        this.isSlotEmpty = true;
    }
    _handleClick() {
        if (this.disabled) {
            return;
        }
        this.checked = !this.checked;
        this.dispatchEvent(new CustomEvent('vsc-change', {
            detail: {
                checked: this.checked,
                label: this.label,
                value: this.value,
            },
            bubbles: true,
            composed: true,
        }));
    }
    _handleKeyDown(event) {
        if (!this.disabled && (event.key === 'Enter' || event.key === ' ')) {
            this.checked = !this.checked;
        }
    }
    _handleSlotChange() {
        this.isSlotEmpty = this.innerHTML === '';
    }
    static get styles() {
        return [super.styles, baseStyles, checkboxStyles, formHelperTextStyles];
    }
    render() {
        const isLabelEmpty = !this.label && this.isSlotEmpty;
        const iconClasses = classMap({
            icon: true,
            checked: this.checked,
            'before-empty-label': isLabelEmpty,
        });
        const labelInnerClasses = classMap({
            'label-inner': true,
            empty: isLabelEmpty,
        });
        const icon = html `<svg
      width="16"
      height="16"
      viewBox="0 0 16 16"
      xmlns="http://www.w3.org/2000/svg"
      fill="currentColor"
    >
      <path
        fill-rule="evenodd"
        clip-rule="evenodd"
        d="M14.431 3.323l-8.47 10-.79-.036-3.35-4.77.818-.574 2.978 4.24 8.051-9.506.764.646z"
      />
    </svg>`;
        const check = this.checked ? icon : nothing;
        return html `
      <div class="wrapper">
        <input
          id="${this._uid}"
          class="checkbox"
          type="checkbox"
          ?checked="${this.checked}"
          value="${this.value}"
          tabindex="-1"
        />
        <div class="${iconClasses}">${check}</div>
        <label for="${this._uid}" class="label" @click="${this._handleClick}">
          <span class="${labelInnerClasses}">
            <slot @slotchange="${this._handleSlotChange}">${this.label}</slot>
          </span>
        </label>
      </div>
    `;
    }
};
__decorate([
    property({ type: Boolean })
], VscodeCheckbox.prototype, "checked", void 0);
__decorate([
    property()
], VscodeCheckbox.prototype, "label", void 0);
__decorate([
    property()
], VscodeCheckbox.prototype, "value", void 0);
__decorate([
    property({ type: Boolean })
], VscodeCheckbox.prototype, "disabled", void 0);
__decorate([
    state()
], VscodeCheckbox.prototype, "isSlotEmpty", void 0);
VscodeCheckbox = __decorate([
    customElement('vscode-checkbox')
], VscodeCheckbox);
export { VscodeCheckbox };
