// @bendera/vscode-webview-elements/dist/vscode-table-row.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { css, html } from 'lit';
import { customElement, property } from 'lit/decorators.js';
import { VscElement } from './includes/VscElement';
let VscodeTableRow = class VscodeTableRow extends VscElement {
    constructor() {
        super(...arguments);
        this.role = 'row';
    }
    static get styles() {
        return [
            super.styles,
            css `
        :host {
          display: table-row;
          width: 100%;
        }

        :host-context(vscode-table[compact]) {
          display: block;
        }

        :host-context(vscode-table[compact][bordered]) {
          border-top: 1px solid var(--vscode-editorGroup-border);
        }
      `,
        ];
    }
    render() {
        return html ` <slot></slot> `;
    }
};
__decorate([
    property({ reflect: true })
], VscodeTableRow.prototype, "role", void 0);
VscodeTableRow = __decorate([
    customElement('vscode-table-row')
], VscodeTableRow);
export { VscodeTableRow };
