// @bendera/vscode-webview-elements/dist/vscode-table-header.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { css, html } from 'lit';
import { customElement, property } from 'lit/decorators.js';
import { VscElement } from './includes/VscElement';
let VscodeTableHeader = class VscodeTableHeader extends VscElement {
    constructor() {
        super(...arguments);
        this.role = 'rowgroup';
    }
    static get styles() {
        return [
            super.styles,
            css `
        :host {
          background-color: rgba(130, 130, 130, 0.04);
          display: table;
          table-layout: fixed;
          width: 100%;
        }
      `,
        ];
    }
    render() {
        return html ` <slot></slot> `;
    }
};
__decorate([
    property({ reflect: true })
], VscodeTableHeader.prototype, "role", void 0);
VscodeTableHeader = __decorate([
    customElement('vscode-table-header')
], VscodeTableHeader);
export { VscodeTableHeader };
