// @bendera/vscode-webview-elements/dist/vscode-radio-group.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { css, html } from 'lit';
import { customElement, property } from 'lit/decorators.js';
import { VscElement } from './includes/VscElement';
let VscodeRadioGroup = class VscodeRadioGroup extends VscElement {
    constructor() {
        super(...arguments);
        this.variant = 'horizontal';
        this.inline = false;
    }
    static get styles() {
        return [
            super.styles,
            css `
        :host {
          display: block;
        }

        .wrapper {
          display: flex;
          flex-wrap: wrap;
        }

        :host([variant='vertical']) .wrapper {
          display: block;
        }

        ::slotted(vscode-radio) {
          margin-right: 20px;
        }

        ::slotted(vscode-radio:last-child) {
          margin-right: 0;
        }

        :host([variant='vertical']) ::slotted(vscode-radio) {
          display: block;
          margin-bottom: 15px;
        }

        :host([variant='vertical']) ::slotted(vscode-radio:last-child) {
          margin-bottom: 0;
        }
      `,
        ];
    }
    render() {
        return html `
      <div class="wrapper">
        <slot></slot>
      </div>
    `;
    }
};
__decorate([
    property({ reflect: true })
], VscodeRadioGroup.prototype, "variant", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], VscodeRadioGroup.prototype, "inline", void 0);
VscodeRadioGroup = __decorate([
    customElement('vscode-radio-group')
], VscodeRadioGroup);
export { VscodeRadioGroup };
