// @bendera/vscode-webview-elements/dist/vscode-form-description.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { css, html } from 'lit';
import { customElement } from 'lit/decorators.js';
import { VscElement } from './includes/VscElement';
let VscodeFormDescription = class VscodeFormDescription extends VscElement {
    static get styles() {
        return [
            super.styles,
            css `
        :host {
          color: var(--vscode-foreground);
          cursor: default;
          display: block;
          font-family: var(--vscode-font-family);
          font-size: var(--vscode-font-size);
          line-height: 1.4;
          margin: 3px 0;
          opacity: 0.9;
          overflow: hidden;
          user-select: text;
        }

        ::slotted(a),
        ::slotted(a:visited) {
          color: var(--vscode-textLink-foreground);
          text-decoration: none;
        }

        ::slotted(a:hover),
        ::slotted(a:active) {
          color: var(--vscode-textLink-activeForeground);
          text-decoration: underline;
        }

        ::slotted(code) {
          color: var(--vscode-textPreformat-foreground);
        }

        ::slotted(p) {
          line-height: 1.4;
          margin-bottom: 9px;
          margin-top: 9px;
        }

        ::slotted(p:first-child) {
          margin-top: 0;
        }

        ::slotted(p:last-child) {
          margin-bottom: 0;
        }
      `,
        ];
    }
    render() {
        return html ` <slot></slot> `;
    }
};
VscodeFormDescription = __decorate([
    customElement('vscode-form-description')
], VscodeFormDescription);
export { VscodeFormDescription };
